### PR TITLE
Add pseudo boolean file parser

### PR DESCRIPTION
**Pull request by freemin7**
Parser is for linear pseudo boolean files only. 
The file has not been compiled and is not integrated in the code base. It passes cpplint. This is my first "production" C++ code

- [ ] Integrate into something something CMAKE or stuff
- [ ] Add to file handling code path
- [ ] Make sure it actually builds
- [ ] Verify objoff has right sign by solving the file 
```
min: 3 ~x1
``` 
and getting objective 0 and 
```
min: 3 ~x1
~x1 >= 1
```
and getting objective 3 and
```
min: 3 x1
x1 >= 1
```
and getting objective 3
- [ ] Make that it so fractional values are allowed (and default) instead of forbidding them 
- [x] Add test instances

I will need some help with 1, 2, 5. If 2 is done but something fails, i am going to fix the code.

### CODE REVIEW

**Comment by ambros-gleixner** [comment]
Two things:
* Can you add a parser test similar to test/papilo/io/MpsParserTest.cpp ? I.e. add a pbo file to test/instances and confirm correct dimensions of the problem. If you want to improve the test, you could even add a check for the actual objective and coefficients.
* Can you make the coding style consistent and run it through clang auto formatter with the .clang-format in the top directory?

**Comment by freemin7** [comment]
Will look into that!

**Comment by freemin7** [comment]
I have not run the tests because i am not sure how the build infrastructure stuff works yet. Added test for the one advanced feature the ~ syntax

**Comment by ambros-gleixner** [comment]
@fschloesser Can you help integrate the tests into the build?

**Comment by fschloesser** [comment]
The new parser tests from `test/papilo/io/MpsParserTest.cpp` should now be running and succeed on my machine

**Comment by freemin7** [comment]
Awesome! That means my mental math worked.
There are still some open ToDos. It is not clear to me if we are blocked on anything i can help with.

**Comment by ambros-gleixner** [comment]
For 2 "add to file handling" I suppose you mean that the PboParser is automatically called on *.pbo files? For this you would need something like

```
   if( boost::algorithm::ends_with( filename, ".pbo" ) )
         prob = PboParser<REAL>::loadProblem( opts.instance_file );
   else
         prob = MpsParser<REAL>::loadProblem( opts.instance_file );
```
in Wrappers.hpp.  But you also want to be called on compressed files, so if .pbo.gz or .pbo.bz2 is the ending. Ideally, you want a while loop stripping .gz and .bz2 from the end and then checking the remaining extension.

For 5 "fractional values" I am not sure what the question is?

**Comment by freemin7** [comment]
> For 5 "fractional values" I am not sure what the question is?

In the PBO world making rounding errors seems to be unacceptable, a sensible default would be fractional presolving. The MPS loader has some code which i interpret to mean using Floating point numbers is required of REAL: https://github.com/scipopt/papilo/blob/71b9cf45b2d92a86fb95de33c250ca656a615164/src/papilo/io/MpsParser.hpp#L61 that is the opposite what i would want.

**Comment by ambros-gleixner** [comment]
Something different: I would be curious what you are using this for.  Maybe we can start a private conversation by mail, do you want to drop me a message at gleixner[AT]zib.de?

File: src/papilo/io/PboParser.hpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                               parser.nRows, true },
+          std::move( parser.rowlhs ), std::move( parser.rowrhs ),
+          std::move( parser.row_flags ), true );
+      problem.setVariableDomains(
+          std::move( Vec<REAL> vect( n, REAL( 0 ) ) ),
+          std::move( Vec<REAL> vect( n, REAL( 1 ) ) ),
+          std::move( Vec<ColFlags> vect( n, kIntegral ) ) );
+      problem.setVariableNames( std::move( parser.colnames ) );
+      problem.setName( std::move( filename ) );
+      // We do not have ConstraintNames in PBO
```
Comment (alexhoen) [inline src/papilo/io/PboParser.hpp:112]: we definitely should give them names. 
Imho a simple counter would be the best solution.

File: src/papilo/io/PboParser.hpp [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+   REAL rhsoff = REAL( 0 );
+
+   std::stringstream row( beginSpace );
+
+   int degree = 0;
+   int variable_index;
+   REAL weight;
+
+   // I am unfamiliar with error handling conventions in this code base.
```
Comment (alexhoen) [inline src/papilo/io/PboParser.hpp:210]: If an error or unexpected data structure is occurring I think is best to abort to signal the user that the code does not support the code 100%

File: src/papilo/io/PboParser.hpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+struct RealParseType<REAL, true>
+{
+   using type = REAL;
+};
+
+/// Parser for pbo files derived from MpsParser.hpp
+template <typename REAL>
+class PboParser
+{
+   static_assert(
```
[the remark this answers is not in the document]
Comment (freemin7) [reply src/papilo/io/PboParser.hpp:77]: I mean rational still are real numbers so this template string makes some sense for me.
If you have a better naming suggestion i welcome it.

File: src/papilo/io/PboParser.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+/*                                                                           */
+/*               This file is part of the program and library                */
+/*    PaPILO --- Parallel Presolve for Integer and Linear Optimization       */
```
[the remark this answers is not in the document]
Comment (freemin7) [reply src/papilo/io/PboParser.hpp:4]: I guess i meant `std:fill` then. I was under the impression that someone managed run the tests which would imply it compiled for them. Apparently not.

I have no idea what i did wrong with the vector constructor there but i will think about it.

**Comment by alexhoen** [comment]
we finished this work in a seperate merge request

**Comment by freemin7** [comment]
I hope my code was at least somewhat useful in that.

**Comment by ambros-gleixner** [comment]
Yes, of course, thanks again!

**Pull request closed without merging**